File: src/main/java/org/snake/models/GameMatrix.java
package org.snake.models;

import org.snake.models.basic.Point;
import org.snake.views.FigureView;
import org.snake.views.GrassView;

import java.util.Arrays;

public class GameMatrix {

    private final int width;
    private final int height;
    private final FigureView[][] matrix;

    public GameMatrix(int width, int height) {
        this.width = width;
        this.height = height;
        this.matrix = new FigureView[width][height];
        for (FigureView[] figureViews : matrix) {
            Arrays.fill(figureViews, GrassView.INSTANCE);
        }
    }

    public FigureView get(int x, int y) {
        return matrix[x][y];
    }

    public FigureView get(Point point) {
        return get(point.x, point.y);
    }

    public void setView(FigureView view, Point position) {
        setView(view, position.x, position.y);
    }

    public void setView(FigureView view, int x, int y) {
        matrix[x][y] = view == null ? GrassView.INSTANCE : view;
    }

    public void removeView(Point point) {
        removeView(point.x, point.y);
    }

    public void removeView(int x, int y) {
        matrix[x][y] = GrassView.INSTANCE;
    }

    public int width() {
        return width;
    }

    public int height() {
        return height;
    }

    public boolean isEmpty(int x, int y) {
        return matrix[x][y].isEmptySpace();
    }

    public boolean isEmpty(Point newPoint) {
        return isEmpty(newPoint.x, newPoint.y);
    }
}

File: src/main/kotlin/org/snake/models/GameMatrix.kt
package org.snake.models

import org.snake.models.basic.Point
import org.snake.views.FigureView
import org.snake.views.GrassView
import java.util.*

class GameMatrix(private val width: Int, private val height: Int) {
    private val matrix: Array<Array<FigureView?>>
    operator fun get(x: Int, y: Int): FigureView? {
        return matrix[x][y]
    }

    operator fun get(point: Point): FigureView? {
        return get(point.x, point.y)
    }

    fun setView(view: FigureView?, position: Point?) {
        setView(view, position!!.x, position.y)
    }

    fun setView(view: FigureView?, x: Int, y: Int) {
        matrix[x][y] = view ?: GrassView
    }

    fun removeView(point: Point?) {
        removeView(point!!.x, point.y)
    }

    fun removeView(x: Int, y: Int) {
        matrix[x][y] = GrassView
    }

    fun width(): Int {
        return width
    }

    fun height(): Int {
        return height
    }

    fun isEmpty(x: Int, y: Int): Boolean {
        return matrix[x][y]!!.isEmptySpace
    }

    fun isEmpty(newPoint: Point?): Boolean {
        return isEmpty(newPoint!!.x, newPoint.y)
    }

    init {
        matrix = Array(width) { arrayOfNulls(height) }
        for (figureViews in matrix) {
            Arrays.fill(figureViews, GrassView)
        }
    }
}